src/sdk/pynni/nni/compression/torch/pruning/structured_pruning.py

```diff
@@ -108,7 +111,156 @@ def get_mask(self, base_mask, weight, num_prune, wrapper, wrapper_idx):
+        """
+        # sparsity configs for each wrapper
+        sparsities = [_w.config['sparsity'] for _w in wrappers]
+        # check the type of the input wrappers
+        for _w in wrappers:
+            msg = 'module type {} is not supported!'.format(_w.type)
+            assert _w.type == 'Conv2d', msg
+        # Among the dependent layers, the layer with smallest
+        # sparsity determines the final benefit of the speedup
+        # module. To better harvest the speed benefit, we need
+        # to ensure that these dependent layers have at least
+        # `min_sparsity` pruned channel are the same.
+        if len(channel_dsets) == len(wrappers):
+            # all the layers in the dependency sets are pruned
+            min_sparsity = min(sparsities)
+        else:
+            # not all the layers in the dependency set
+            # are pruned
+            min_sparsity = 0
+        # donnot prune the channels that we cannot harvest the speed from
+        sparsities = [min_sparsity] * len(sparsities)
+        # find the max number of the filter groups of the dependent
+        # layers. The group constraint of this dependency set is decided
+        # by the layer with the max groups.
+        max_group = max(groups)
+        channel_count = wrappers[0].module.weight.data.size(0)
+        device = wrappers[0].module.weight.device
+        channel_sum = torch.zeros(channel_count).to(device)
+        for _w, _w_idx in zip(wrappers, wrappers_idx):
+            # calculate the L1/L2 sum for all channels
+            c_sum = self._get_channel_sum(_w, _w_idx)
+
+            if c_sum is None:
+                # if the channel sum cannot be calculated
+                # now, return None
+                return None
+            channel_sum += c_sum
+
+        # prune the same `min_sparsity` channels based on channel_sum
+        # for all the layers in the channel sparsity
+        target_pruned = int(channel_count * min_sparsity)
+        # pruned_per_group may be zero, for example dw conv
+        pruned_per_group = int(target_pruned / max_group)
+        group_step = int(channel_count / max_group)
+
+        channel_masks = []
+        for gid in range(max_group):
+            _start = gid * group_step
+            _end = (gid + 1) * group_step
+            if pruned_per_group > 0:
+                threshold = torch.topk(
+                    channel_sum[_start: _end], pruned_per_group, largest=False)[0].max()
+                group_mask = torch.gt(channel_sum[_start:_end], threshold)
+            else:
+                group_mask = torch.ones(group_step).to(device)
+            channel_masks.append(group_mask)
+        channel_masks = torch.cat(channel_masks, dim=0)
+        pruned_channel_index = (
+            channel_masks == False).nonzero().squeeze(1).tolist()
+        logger.info('Prune the %s channels for all dependent',
+                    ','.join([str(x) for x in pruned_channel_index]))
```

Comment:
this function is too long, suggest to move this conv specific logic into a dedicated member function